### PR TITLE
etcdserver: not processing messages until server is started

### LINKED ISSUE #2705: etcd panic on invalid memory address when RecvAppendReq
```
2015/04/17 19:27:40 etcdserver: heartbeat = 100ms
2015/04/17 19:27:40 etcdserver: election = 1000ms
2015/04/17 19:27:40 etcdserver: snapshot count = 10000
2015/04/17 19:27:40 etcdserver: advertise client URLs = http://10.240.172.103:2379
2015/04/17 19:27:40 etcdserver: loaded cluster information from store: etcd-0=http://10.240.190.54:2380,etcd-1=http://10.240.172.103:2380,etcd-2=http://10.240.21.79:2380
2015/04/17 19:27:40 etcdserver: restart member a9b0ebadefc94b6f in cluster 7a25f25b7fd507b0 at commit index 4166825
2015/04/17 19:27:40 INFO: raft: a9b0ebadefc94b6f became follower at term 232
2015/04/17 19:27:40 INFO: raft: newRaft a9b0ebadefc94b6f [peers: [4f8b3f8e6b7b,8d448318f55e9d1d,a9b0ebadefc94b6f], term: 232, commit: 4166825, applied: 4145644, lastindex: 4167097, lastterm: 232]
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x6235f5]

goroutine 38 [running]:
github.com/coreos/etcd/etcdserver/stats.(*statsQueue).Insert(0x0, 0xc20f686540)
        /Users/unihorn/gopath/src/github.com/coreos/etcd/etcdserver/stats/queue.go:63 +0x1f5
github.com/coreos/etcd/etcdserver/stats.(*ServerStats).RecvAppendReq(0xc20f576480, 0xc20b8988e0, 0xc, 0x39eb4)
        /Users/unihorn/gopath/src/github.com/coreos/etcd/etcdserver/stats/server.go:115 +0x16a
github.com/coreos/etcd/etcdserver.(*EtcdServer).Process(0xc2080b67e0, 0x7f4ea9c46ef8, 0xc208032348, 0x3, 0xa9b0ebadefc94b6f, 0x4f8b3f8e6b7b, 0xeb, 0xe8, 0x3f959d, 0xc2177d0000, ...)
        /Users/unihorn/gopath/src/github.com/coreos/etcd/etcdserver/server.go:335 +0x302
github.com/coreos/etcd/rafthttp.func·002()
        /Users/unihorn/gopath/src/github.com/coreos/etcd/rafthttp/peer.go:175 +0x803
created by github.com/coreos/etcd/rafthttp.startPeer
        /Users/unihorn/gopath/src/github.com/coreos/etcd/rafthttp/peer.go:195 +0x970
```

### PR DESCRIPTION
**Pull request by yichengq**
It fixes a bug that etcdserver may panic because it processes a message
before it does initialization.

fixes #2705

### CODE REVIEW

File: etcdserver/server.go [hunk 1]
```diff
@@ -326,6 +332,13 @@ func (s *EtcdServer) ID() types.ID { return s.id }
 func (s *EtcdServer) RaftHandler() http.Handler { return s.r.transport.Handler() }
 
 func (s *EtcdServer) Process(ctx context.Context, m raftpb.Message) error {
+	s.mu.Lock()
+	started := s.started
+	s.mu.Unlock()
+	if !started {
+		log.Printf("etcdserver: drop message from member %s because etcdserver has not been started", types.ID(m.From))
```
Comment (xiang90) [inline etcdserver/server.go]: i do not think the error message is useful to the end users. this is more like a very uncommon internal race.

**Comment by xiang90** [comment]
How about making `start` a `startc` chan. And install the http handler after i am sure the server is started.

**Comment by yichengq** [comment]
The message is received from streamReader. The details of steps are as follows:
1. call NewServer
2. call newTransport, and set Server as the message processor of transport
3. AddMember into transport
4. the new member builds the read stream to the remote, and receive message from the remote
5. Server receives message, but has not been inited.

http handler part is fine. We init the server before installing http handler.

I am finding a better way to fix it, considering this way is not straight-forward.

**Comment by xiang90** [comment]
> We init the server before installing http handler.

We do not wait for the server to be started though.

**Comment by yichengq** [comment]
@xiang90 We install http handler after initializing the server internal structs, but not after starting for loop.

**Pull request closed without merging**